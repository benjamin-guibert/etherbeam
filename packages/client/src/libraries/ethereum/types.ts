import { BigNumber } from 'ethers'

export const WETH_HASH = '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2'
export const USDC_HASH = '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48'

export enum AddressType {
  Unknown = 'unknown',
  Wallet = 'wallet',
  Token = 'token',
  Contract = 'contract',
  Transaction = 'transaction',
}

export enum TransactionMethodParameterType {
  Unknown = 'unknown',
  UnsignedInteger = 'unsigned_integer',
  Address = 'address',
  Addresses = 'addresses',
}

export enum TransactionStatus {
  Pending = 'pending',
  Mined = 'mined',
  Validated = 'validated',
  Cancelled = 'cancelled',
}

export enum TransactionActionType {
  Unknown = 'unknown',
  Approval = 'approval',
  Transfer = 'transfer',
  Swap = 'swap',
}

export enum TransactionActionDirection {
  Buy = 'buy',
  Sell = 'sell',
}

export interface TransactionMethodParameter {
  name: string
  type: TransactionMethodParameterType
  rawType: string
  value: TransactionMethodValueType
}

export enum HistoryTime {
  Hour = 'h',
  Day = 'd',
  Week = 'w',
  Month = 'm',
  Year = 'y',
}

export interface TokenPriceHistory {
  time: HistoryTime
  price: BigNumber
  ratio: number
}

export interface Token extends Address {
  name: string
  description?: string
  symbol: string
  decimals: number
  chartPair?: string
  price?: BigNumber
  priceHistory: TokenPriceHistory[]
  website?: string
  whitepaper?: string
  github?: string
  linkedin?: string
  facebook?: string
  reddit?: string
  twitter?: string
  telegram?: string
  discord?: string
  actions?: TransactionAction[]
}

export interface Address {
  hash: string
  url: string
  type: AddressType
  label?: string
}

export interface Transaction {
  blockNumber?: number
  dateTime: Date
  address: Address
  status: TransactionStatus
  from?: Address
  to?: Address
  value?: BigNumber
  gas: {
    unitPrice?: BigNumber
    limit?: number
    used?: number
    ratio?: number
    price?: BigNumber
  }
  action?: TransactionMethod
  logs?: TransactionMethod[]
}

export interface TransactionMethod {
  contract: Address
  name: string
  parameters: TransactionMethodParameter[]
}

export type TransactionMethodValueType = string | BigNumber | Address | Address[]

export interface TransactionAction {
  index: number
  transaction: Transaction
  type: TransactionActionType
  direction: TransactionActionDirection
  holder: Address
  fromAddress?: Address | Token
  fromAmount?: BigNumber
  toAddress?: Address | Token
  toAmount?: BigNumber
}
