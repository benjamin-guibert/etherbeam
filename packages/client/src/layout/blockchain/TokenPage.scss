@use '../../styles/resources/sizes';
@use '../../styles/resources/fonts';
@use '../../styles/resources/colors';

.my-tokenpage {
  &-subheader {
    justify-content: space-between;
  }

  &-links li {
    padding-left: 0px !important;
  }

  &-price {
    color: colors.$fg-disabled;
    font-size: sizes.$size-l;
  }

  &-history {
    margin: sizes.$space-s 0px !important;
  }
}
