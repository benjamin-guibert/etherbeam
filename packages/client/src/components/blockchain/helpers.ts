import { faCoins, faFile, faWallet, IconDefinition } from '@fortawesome/free-solid-svg-icons'
import { BigNumber, ethers } from 'ethers'
import numeral from 'numeral'
import { AddressType, Token } from '../../libraries/ethereum/types'

const ETH_DECIMALS = 18
const MAX_DECIMALS = 13

export interface AmountPrintOptions {
  forceDecimals?: boolean
  decimals?: number
}

const printAmount = (amount: string, decimals: number, options?: AmountPrintOptions): string => {
  const { forceDecimals, decimals: forcedDecimals } = options ?? {}
  const decimalsCount = Math.min(forcedDecimals || decimals, MAX_DECIMALS)
  const decimalFormat = forceDecimals ? '0'.repeat(decimalsCount) : `[${'0'.repeat(decimalsCount)}]`

  return amount ? numeral(amount).format(`0,0.${decimalFormat}`) : '?'
}

export type Unit = 'wei' | 'kwei' | 'mwei' | 'gwei' | 'ether'

export const printEtherAmount = (amount: BigNumber, options?: AmountPrintOptions, unit?: Unit): string =>
  printAmount(amount ? ethers.utils.formatUnits(amount, unit) : null, ETH_DECIMALS, options)

export const printTokenAmount = (amount: BigNumber, token: Token, options?: AmountPrintOptions): string => {
  const { decimals } = token

  return printAmount(amount ? ethers.utils.formatUnits(amount, decimals) : null, decimals, options)
}

export const shortenAddress = (address: string): string => {
  const start = address.substring(0, 6)
  const end = address.substring(address.length - 4)

  return `${start}...${end}`
}

export const getAddressTypeIcon = (type: AddressType): IconDefinition => {
  switch (type) {
    case AddressType.Wallet:
      return faWallet
    case AddressType.Contract:
      return faFile
    case AddressType.Token:
      return faCoins
    default:
      return null
  }
}
