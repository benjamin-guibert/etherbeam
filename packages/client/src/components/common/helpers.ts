import { faDotCircle, faFile, IconDefinition } from '@fortawesome/free-solid-svg-icons'
import numeral from 'numeral'
import { AddressType, Token } from 'libraries/ethereum/types'
import { BigNumber, ethers } from 'ethers'

const ETH_SYMBOL = 'ETH'
const ETH_DECIMALS = 18

export interface AmountPrintOptions {
  forceDecimals?: boolean
  decimals?: number
  withSymbol?: boolean
}

const printAmount = (amount: string, symbol: string, decimals: number, options?: AmountPrintOptions): string => {
  const { forceDecimals, decimals: forcedDecimals, withSymbol } = options ?? {}
  const decimalsCount = forcedDecimals ?? decimals
  const decimalFormat = forceDecimals ? '0'.repeat(decimalsCount) : `[${'0'.repeat(decimalsCount)}]`
  const printedAmount = amount ? numeral(amount).format(`0,0.${decimalFormat}`) : '?'

  return `${withSymbol ? `${symbol} ` : ''}${printedAmount}`
}

export type Unit = 'wei' | 'kwei' | 'mwei' | 'gwei' | 'ether'

export const printEtherAmount = (amount: BigNumber, options?: AmountPrintOptions, unit?: Unit): string =>
  printAmount(amount ? ethers.utils.formatUnits(amount, unit) : null, ETH_SYMBOL, ETH_DECIMALS, options)

export const printTokenAmount = (amount: BigNumber, token: Token, options?: AmountPrintOptions): string => {
  const { symbol, decimals } = token

  return printAmount(amount ? ethers.utils.formatUnits(amount, decimals) : null, symbol, decimals, options)
}

export const shortenAddress = (address: string): string => {
  const start = address.substring(0, 6)
  const end = address.substring(address.length - 4)

  return `${start}...${end}`
}

export const getAddressTypeIcon = (type: AddressType): IconDefinition => {
  switch (type) {
    case AddressType.Contract:
      return faFile
    case AddressType.Token:
      return faDotCircle
    default:
      return null
  }
}
