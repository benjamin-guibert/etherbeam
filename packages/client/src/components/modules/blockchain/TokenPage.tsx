import React, { ReactElement, useState, useRef, useEffect } from 'react'
import { useParams } from 'react-router-dom'
import Alert from 'react-bootstrap/Alert'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'
import Loader from 'components/common/Loader'
import Address from 'components/common/Address'
import { ToastContent } from 'components/Toaster'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faHome, faFileAlt, IconDefinition } from '@fortawesome/free-solid-svg-icons'
import {
  faGithub,
  faLinkedin,
  faFacebook,
  faReddit,
  faTwitter,
  faTelegram,
  faDiscord,
} from '@fortawesome/free-brands-svg-icons'
import IconLink from 'components/common/IconLink'
import { printEtherAmount } from 'components/common/helpers'
import Difference from 'components/common/Difference'
import { Token, TransactionActionType } from 'libraries/ethereum/types'
import { fetchToken } from 'libraries/ethereum/server'
import Page from 'components/common/Page'
import TransactionActionsList from './TransactionActionsList'

const FETCH_TOKEN_INTERVAL = 30 * 1000

interface TokenPageParams {
  address: string
}

interface TokenPageProps {
  addToast: (toast: ToastContent) => void
}

const Link = ({ url, title, icon }: { url: string; title: string; icon: IconDefinition }): ReactElement => {
  return (
    <IconLink url={url} title={title}>
      <FontAwesomeIcon icon={icon} size="lg" />
    </IconLink>
  )
}

const TokenPage = ({ addToast }: TokenPageProps): ReactElement => {
  const [token, setToken] = useState<Token>(undefined)
  const [alert, setAlert] = useState<string>(null)
  const addToastRef = useRef(addToast)
  const addressParam = useParams<TokenPageParams>().address
  const {
    name,
    symbol,
    website,
    whitepaper,
    github,
    linkedin,
    facebook,
    reddit,
    twitter,
    telegram,
    discord,
    price,
    priceHistory,
    actions,
  } = token || {}
  const filteredActions = actions?.filter((action) => action.type != TransactionActionType.Approval)

  useEffect(() => {
    addToastRef.current = addToast
  }, [addToast])

  useEffect(() => {
    const getToken = async (): Promise<void> => {
      fetchToken(addressParam, 'actions')
        .then((fetchedToken) => {
          if (!fetchedToken) return setAlert(`The token '${addressParam}' is unknown.`)

          setToken(fetchedToken)
        })
        .catch(() => addToastRef.current({ type: 'error', message: 'Error while fetching token.' }))
    }
    const startFetchingToken = async (): Promise<void> => {
      await getToken()
    }

    const interval = setInterval(getToken, FETCH_TOKEN_INTERVAL)

    startFetchingToken()

    return () => {
      clearInterval(interval)
    }
  }, [addressParam])

  return (
    <>
      {!!token && !alert && (
        <Page title={name}>
          <>
            <Row>
              <Col className="my-auto d-none d-lg-inline">
                <Address address={token} size="m" noFlag />
              </Col>
              <Col className="my-auto d-none d-sm-inline d-lg-none">
                <Address address={token} size="m" noFlag short />
              </Col>
              <Col className="ml-auto text-sm-right">
                {website && <Link url={website} title="Website" icon={faHome} />}
                {whitepaper && <Link url={whitepaper} title="White paper" icon={faFileAlt} />}
                {github && <Link url={github} title="Github" icon={faGithub} />}
                {linkedin && <Link url={linkedin} title="LinkedIn" icon={faLinkedin} />}
                {facebook && <Link url={facebook} title="Facebook" icon={faFacebook} />}
                {reddit && <Link url={reddit} title="Reddit" icon={faReddit} />}
                {twitter && <Link url={twitter} title="Twitter" icon={faTwitter} />}
                {telegram && <Link url={telegram} title="Telegram" icon={faTelegram} />}
                {discord && <Link url={discord} title="Discord" icon={faDiscord} />}
              </Col>
            </Row>
            {price && (
              <div className="app-text-lg my-3">
                <Row>
                  <Col className="text-monospace">
                    <span className="d-none d-sm-inline text-muted">{symbol} 1 = </span>
                    {printEtherAmount(price, { withSymbol: true })}
                  </Col>
                </Row>
                {!!priceHistory.length && (
                  <Row>
                    {priceHistory.map(({ time, ratio }, index) => (
                      <Col className="my-auto" key={index}>
                        <Difference time={time} ratio={ratio} />
                      </Col>
                    ))}
                  </Row>
                )}
              </div>
            )}
            {!!filteredActions?.length && <TransactionActionsList actions={filteredActions} />}
          </>
        </Page>
      )}
      {!token && !alert && <Loader />}
      {!!alert && <Alert variant="danger">{alert}</Alert>}
    </>
  )
}

export default TokenPage
