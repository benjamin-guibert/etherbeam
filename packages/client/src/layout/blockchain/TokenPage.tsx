import React, { FC } from 'react'
import { faHome, faFileAlt, IconDefinition, faPlusCircle, faMinusCircle } from '@fortawesome/free-solid-svg-icons'
import {
  faGithub,
  faLinkedin,
  faFacebook,
  faReddit,
  faTwitter,
  faTelegram,
  faDiscord,
} from '@fortawesome/free-brands-svg-icons'
import { Token, TransactionActionType } from '../../libraries/ethereum/types'
import { getUniswapTradeUrl } from '../../components/blockchain/helpers'
import Alert from '../../components/Alert'
import Link from '../../components/Link'
import Loader from '../../components/Loader'
import List from '../../components/List'
import ListItem from '../../components/ListItem'
import AddressItem from '../../components/blockchain/AddressItem'
import TokenAmount from '../../components/blockchain/TokenAmount'
import TokenPriceHistory from '../../components/blockchain/TokenPriceHistory'
import TokenChart from '../../components/blockchain/TokenChart'
import TransactionActionList from '../../components/blockchain/TransactionActionList'
import './TokenPage.scss'

interface IconLinkProps {
  url: string
  icon: IconDefinition
  description: string
}

const IconLink: FC<IconLinkProps> = ({ url, icon, description }) => {
  return (
    <ListItem>
      <Link href={url} icon={icon} description={description} blank />
    </ListItem>
  )
}

interface TokenPageProps {
  token: Token
  alert?: string
  loading?: boolean
}

const TokenPage: FC<TokenPageProps> = ({ token, alert, loading }) => {
  const {
    hash,
    name,
    symbol,
    chartPair,
    website,
    whitepaper,
    github,
    linkedin,
    facebook,
    reddit,
    twitter,
    telegram,
    discord,
    price,
    priceHistory,
    actions,
  } = token || {}
  const filteredActions = actions?.filter((action) => action.type != TransactionActionType.Approval)

  return (
    <>
      <h1>
        {token && name}
        {!token && loading && <Loader type="dots" />}
        {!token && !loading && !!alert && 'Unknown token'}
      </h1>

      {token && (
        <>
          <h2 className="my-tokenpage-subheader">
            <List horizontal className="my-tokenpage-links">
              {website && <IconLink url={website} description="Website" icon={faHome} />}
              {whitepaper && <IconLink url={whitepaper} description="Whitepaper" icon={faFileAlt} />}
              {github && <IconLink url={github} description="Github" icon={faGithub} />}
              {linkedin && <IconLink url={linkedin} description="LinkedIn" icon={faLinkedin} />}
              {facebook && <IconLink url={facebook} description="Facebook" icon={faFacebook} />}
              {reddit && <IconLink url={reddit} description="Reddit" icon={faReddit} />}
              {twitter && <IconLink url={twitter} description="Twitter" icon={faTwitter} />}
              {telegram && <IconLink url={telegram} description="Telegram" icon={faTelegram} />}
              {discord && <IconLink url={discord} description="Discord" icon={faDiscord} />}
            </List>
            <>
              <AddressItem className="my-d-min-m" address={token} noFlag />
              <AddressItem className="my-d-max-m" address={token} noFlag short />
            </>
          </h2>
          <section>
            <span className="my-d-min-xs my-disabled-fg my-size-l">
              <span className="my-symbol">{symbol}</span>
              <span className="my-amount"> 1 = </span>
            </span>
            <TokenAmount amount={price} size="l" />
            {!!priceHistory?.length && (
              <TokenPriceHistory priceHistory={priceHistory} size="m" className="my-tokenpage-history" />
            )}
          </section>
          {chartPair && (
            <section>
              <TokenChart pair={chartPair} height={300} />
            </section>
          )}
          <section>
            <List horizontal center>
              <ListItem>
                <Link
                  href={getUniswapTradeUrl(null, hash)}
                  icon={faPlusCircle}
                  label="Buy"
                  color="positive"
                  size="l"
                  button
                  blank
                />
              </ListItem>
              <ListItem>
                <Link
                  href={getUniswapTradeUrl(hash, null)}
                  icon={faMinusCircle}
                  label="Sell"
                  color="negative"
                  size="l"
                  button
                  blank
                />
              </ListItem>
            </List>
          </section>
          {!!filteredActions?.length && (
            <section>
              <h3>Actions</h3>
              <TransactionActionList actions={filteredActions} />
            </section>
          )}
        </>
      )}
      {!token && !!alert && <Alert color="negative">{alert}</Alert>}
    </>
  )
}

export default TokenPage
