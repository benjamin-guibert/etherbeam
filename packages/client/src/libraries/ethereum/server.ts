import axios from 'axios'
import compact from 'lodash/compact'
import { HttpStatus, getPagination, PaginatedContent } from 'libraries/api'
import { Token, Transaction, USDC_HASH, WETH_HASH } from './types'
import { TokenData, TransactionData } from './serverTypes'
import { parseTokenData, parseTransactionData } from './serverParsers'

const TIMEOUT = 60 * 1000
const TRANSACTIONS_PER_PAGE = 20

const axiosClient = axios.create({
  baseURL: process.env.SERVER_API_HOST,
  headers: { 'Content-Type': 'application/json', Accept: 'application/json' },
  timeout: TIMEOUT,
})

export const fetchTokens = async (): Promise<Token[]> => {
  const { data } = await axiosClient.get<TokenData[]>('tokens')
  const ignoredHashes = [WETH_HASH, USDC_HASH]

  return compact(data.map((token) => (!ignoredHashes.includes(token.address_hash) ? parseTokenData(token) : null)))
}

export const fetchToken = async (address: string, withList?: 'actions'): Promise<Token> => {
  const config = {
    validateStatus: (status: number) => [HttpStatus.Ok, HttpStatus.NotFound].includes(status),
  }

  const { data } = await axiosClient.get<TokenData>(`tokens/${address}?list=${withList || ''}`, config)

  return data ? parseTokenData(data) : null
}

export const fetchPendingTransactions = async (page: number): Promise<PaginatedContent<Transaction>> => {
  const response = await axiosClient.get<TransactionData[]>(
    `block_transactions/pending?items=${TRANSACTIONS_PER_PAGE}&page=${page}`
  )

  return {
    items: response.data.map((transaction) => parseTransactionData(transaction)),
    pagination: getPagination(response),
  }
}
