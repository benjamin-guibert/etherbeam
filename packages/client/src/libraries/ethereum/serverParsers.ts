import { BigNumber } from 'ethers'
import {
  TransactionData,
  TokenData,
  AddressData,
  TokenActionData,
  TransactionMethodParameterData,
  TransactionMethodData,
} from './serverTypes'
import {
  Transaction,
  Token,
  HistoryTime,
  Address,
  AddressType,
  TransactionMethodValueType,
  TransactionAction,
  TransactionActionDirection,
  TransactionMethodParameter,
  TransactionMethodParameterType,
  TransactionMethod,
} from './types'
import { createTransactionAddress, getEtherscanUrl } from './ethereum'
import orderBy from 'lodash/orderBy'

const parseBigNumberFromServer = (value: string): BigNumber => (value ? BigNumber.from(value.split('.')[0]) : null)

export const parseAddressData = ({ address_hash, address_type, label }: AddressData): Address => {
  return {
    hash: address_hash,
    type: address_type,
    label,
    url: getEtherscanUrl('address', address_hash),
  }
}

const parseHashData = (hash: string, address?: AddressData | TokenData): Address => {
  if (!address) {
    return {
      hash,
      type: AddressType.Unknown,
      url: getEtherscanUrl('address', hash),
    }
  }

  switch (address.address_type) {
    case AddressType.Token:
      return parseTokenData(address as TokenData)
    default:
      return parseAddressData(address)
  }
}

const parseTransactionMethodValueType = ({
  parameter_type,
  decimal_value,
  addresses,
}: TransactionMethodParameterData): TransactionMethodValueType => {
  switch (parameter_type) {
    case TransactionMethodParameterType.UnsignedInteger:
      return parseBigNumberFromServer(decimal_value)
    case TransactionMethodParameterType.Address:
      return addresses.length > 0 ? parseHashData(addresses[0].address_hash, addresses[0].address) : null
    case TransactionMethodParameterType.Addresses:
      return addresses ? addresses.map((address) => parseHashData(address.address_hash, address.address)) : null
    default:
      return null
  }
}

const parseTransactionMethodParameter = (parameter: TransactionMethodParameterData): TransactionMethodParameter => {
  const { name, parameter_type, raw_type } = parameter

  return {
    name,
    type: parameter_type,
    rawType: raw_type,
    value: parseTransactionMethodValueType(parameter),
  }
}

const parseTransactionMethodData = ({ contract, name, parameters }: TransactionMethodData): TransactionMethod => {
  return {
    contract: parseAddressData(contract),
    name,
    parameters: parameters.map((parameter) => parseTransactionMethodParameter(parameter)),
  }
}

export const parseTransactionData = ({
  transaction_hash,
  status,
  datetime,
  block_number,
  from_address_hash,
  from_address,
  to_address_hash,
  to_address,
  value,
  gas_unit_price,
  gas_limit,
  gas_used,
  gas_ratio,
  gas_fee,
  transaction_method_action,
  transaction_method_logs,
}: TransactionData): Transaction => {
  return {
    blockNumber: block_number,
    dateTime: datetime,
    address: createTransactionAddress(transaction_hash),
    status,
    from: from_address_hash ? parseHashData(from_address_hash, from_address) : null,
    to: to_address_hash ? parseHashData(to_address_hash, to_address) : null,
    value: parseBigNumberFromServer(value),
    gas: {
      unitPrice: parseBigNumberFromServer(gas_unit_price),
      limit: gas_limit,
      used: gas_used,
      ratio: gas_ratio,
      price: parseBigNumberFromServer(gas_fee),
    },
    action: transaction_method_action ? parseTransactionMethodData(transaction_method_action) : null,
    logs: transaction_method_logs?.map((log) => parseTransactionMethodData(log)),
  }
}

const parseTokenActionData = (
  {
    block_transaction,
    holder_address_hash,
    holder_address,
    index,
    action_type,
    from_address,
    from_address_hash,
    from_amount,
    to_address,
    to_address_hash,
    to_amount,
  }: TokenActionData,
  direction: TransactionActionDirection
): TransactionAction => {
  return {
    transaction: parseTransactionData(block_transaction),
    type: action_type,
    direction,
    index,
    holder: parseHashData(holder_address_hash, holder_address),
    fromAddress: from_address_hash ? parseHashData(from_address_hash, from_address) : null,
    fromAmount: parseBigNumberFromServer(from_amount),
    toAddress: to_address_hash ? parseHashData(to_address_hash, to_address) : null,
    toAmount: parseBigNumberFromServer(to_amount),
  }
}

export const parseTokenData = (data: TokenData): Token => {
  const {
    address_hash,
    name,
    description,
    symbol,
    decimals,
    price,
    price_1h,
    price_1h_ratio,
    price_1d,
    price_1d_ratio,
    price_1w,
    price_1w_ratio,
    price_1m,
    price_1m_ratio,
    price_1y,
    price_1y_ratio,
    website,
    whitepaper,
    github,
    linkedin,
    facebook,
    reddit,
    twitter,
    telegram,
    discord,
    from_transaction_actions,
    to_transaction_actions,
  } = data
  const priceHistory = []
  if (price_1h) {
    priceHistory.push({ time: HistoryTime.Hour, price: parseBigNumberFromServer(price_1h), ratio: price_1h_ratio })
  }
  if (price_1d) {
    priceHistory.push({ time: HistoryTime.Day, price: parseBigNumberFromServer(price_1d), ratio: price_1d_ratio })
  }
  if (price_1w) {
    priceHistory.push({ time: HistoryTime.Week, price: parseBigNumberFromServer(price_1w), ratio: price_1w_ratio })
  }
  if (price_1m) {
    priceHistory.push({ time: HistoryTime.Month, price: parseBigNumberFromServer(price_1m), ratio: price_1m_ratio })
  }
  if (price_1y) {
    priceHistory.push({ time: HistoryTime.Year, price: parseBigNumberFromServer(price_1y), ratio: price_1y_ratio })
  }

  const fromActions = from_transaction_actions
    ? from_transaction_actions.map((action) => parseTokenActionData(action, TransactionActionDirection.Sell))
    : []
  const toActions = to_transaction_actions
    ? to_transaction_actions.map((action) => parseTokenActionData(action, TransactionActionDirection.Buy))
    : []
  const actions =
    from_transaction_actions || to_transaction_actions
      ? orderBy([...fromActions, ...toActions], ['transaction.dateTime', 'index'], ['desc', 'desc'])
      : null

  return {
    ...parseAddressData(data),
    type: AddressType.Token,
    url: getEtherscanUrl('token', address_hash),
    name,
    description,
    symbol,
    decimals,
    price: price ? parseBigNumberFromServer(price) : null,
    priceHistory,
    website,
    whitepaper,
    github,
    linkedin,
    facebook,
    reddit,
    twitter,
    telegram,
    discord,
    actions,
  }
}
