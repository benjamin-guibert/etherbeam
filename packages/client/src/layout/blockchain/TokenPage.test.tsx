import * as React from 'react'
import { mount, ReactWrapper } from 'enzyme'
import { mocked } from 'ts-jest/utils'
import { act } from 'react-dom/test-utils'
import { MemoryRouter, Route } from 'react-router'
import { clearMocks } from '../../../tests/helpers'
import { createToken } from '../../../tests/fixtures/ethereum'

jest.mock('libraries/ethereum/server')
jest.mock('components/blockchain/TokenChart')

import { fetchToken } from 'libraries/ethereum/server'

const fetchTokenMock = mocked(fetchToken)

import TokenPage from './TokenPage'

describe('<TokenPage />', () => {
  const token = createToken('0x0000000000000000000000000000000000000111')

  let component: ReactWrapper

  describe('Default', () => {
    beforeAll(async () => {
      clearMocks()

      fetchTokenMock.mockResolvedValue(token)

      await act(async () => {
        component = mount(
          <MemoryRouter initialEntries={['/tokens/0x0000000000000000000000000000000000000111']}>
            <Route path="/tokens/:address" exact>
              <TokenPage />
            </Route>
          </MemoryRouter>
        )
        await new Promise(setImmediate)
        component.update()
      })
    })

    it('should fetch tokens', () => expect(fetchTokenMock).toBeCalledTimes(1))

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('Default', () => {
    beforeAll(async () => {
      clearMocks()

      fetchTokenMock.mockResolvedValue(null)

      await act(async () => {
        component = mount(
          <MemoryRouter initialEntries={['/tokens/0x0000000000000000000000000000000000000111']}>
            <Route path="/tokens/:address" exact>
              <TokenPage />
            </Route>
          </MemoryRouter>
        )
        await new Promise(setImmediate)
        component.update()
      })
    })

    it('should fetch tokens', () => expect(fetchTokenMock).toBeCalledTimes(1))

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })
})
