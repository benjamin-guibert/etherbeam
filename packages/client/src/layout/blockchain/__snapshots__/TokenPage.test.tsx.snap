// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TokenPage /> Default should match snapshot 1`] = `
Array [
  <h1>
    Token
  </h1>,
  <h2
    class="my-tokenpage-subheader"
  >
    <ul
      class="my-list-h  my-tokenpage-links"
    />
    <span
      class="my-d-min-m"
    >
      <a
        class="my-clickable  my-link my-secondary-fg"
        href="https://etherscan.io/address/0x0000000000000000000000000000000000000111"
        tabindex="0"
        target="_blank"
      >
        <span
          class="my-addresshash"
        >
          0x0000000000000000000000000000000000000111
        </span>
      </a>
    </span>
    <span
      class="my-d-max-m"
    >
      <a
        class="my-clickable  my-link my-secondary-fg"
        href="https://etherscan.io/address/0x0000000000000000000000000000000000000111"
        tabindex="0"
        target="_blank"
      >
        <span
          class="my-addresshash"
        >
          0x0000...0111
        </span>
      </a>
    </span>
  </h2>,
  <section>
    <span
      class="my-d-min-xs my-disabled-fg my-size-l"
    >
      <span
        class="my-symbol"
      >
        TKN
      </span>
      <span
        class="my-amount"
      >
         1 = 
      </span>
    </span>
    <span
      class="my-badge my-light-bg my-size-l "
    >
      <span
        class="my-tokenamount-symbol"
      >
        ETH
      </span>
      <span
        class="my-tokenamount-amount"
      >
         NaN
      </span>
    </span>
    <ul
      class="my-list-h  my-tokenpage-history"
    >
      <li
        class="my-listitem "
      >
        <span
          class="my-badge my-positive-bg my-size-m "
        >
          <span
            class="my-difference-prefix"
          >
            H
          </span>
          <span
            class="my-difference-amount"
          >
            +100%
          </span>
        </span>
      </li>
      <li
        class="my-listitem "
      >
        <span
          class="my-badge my-negative-bg my-size-m "
        >
          <span
            class="my-difference-prefix"
          >
            D
          </span>
          <span
            class="my-difference-amount"
          >
            -50%
          </span>
        </span>
      </li>
      <li
        class="my-listitem "
      >
        <span
          class="my-badge my-negative-bg my-size-m "
        >
          <span
            class="my-difference-prefix"
          >
            W
          </span>
          <span
            class="my-difference-amount"
          >
            -75%
          </span>
        </span>
      </li>
      <li
        class="my-listitem "
      >
        <span
          class="my-badge my-light-bg my-size-m "
        >
          <span
            class="my-difference-prefix"
          >
            M
          </span>
          <span
            class="my-difference-amount"
          >
            +0%
          </span>
        </span>
      </li>
      <li
        class="my-listitem "
      >
        <span
          class="my-badge my-positive-bg my-size-m "
        >
          <span
            class="my-difference-prefix"
          >
            Y
          </span>
          <span
            class="my-difference-amount"
          >
            +300%
          </span>
        </span>
      </li>
    </ul>
  </section>,
  <section>
    <ul
      class="my-list-h my-list-center "
    >
      <li
        class="my-listitem "
      >
        <a
          class="my-clickable  my-link-button my-positive-bg my-size-l"
          href="https://app.uniswap.org/#/swap?outputCurrency=0x0000000000000000000000000000000000000111"
          tabindex="0"
          target="_blank"
        >
          <span
            class="my-icon  "
          >
            <span
              class=" my-alt"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-plus-circle fa-w-16 "
                data-icon="plus-circle"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <span
              class="my-icon-label "
            >
              Buy
            </span>
          </span>
        </a>
      </li>
      <li
        class="my-listitem "
      >
        <a
          class="my-clickable  my-link-button my-negative-bg my-size-l"
          href="https://app.uniswap.org/#/swap?inputCurrency=0x0000000000000000000000000000000000000111&"
          tabindex="0"
          target="_blank"
        >
          <span
            class="my-icon  "
          >
            <span
              class=" my-alt"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-minus-circle fa-w-16 "
                data-icon="minus-circle"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zM124 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H124z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <span
              class="my-icon-label "
            >
              Sell
            </span>
          </span>
        </a>
      </li>
    </ul>
  </section>,
]
`;

exports[`<TokenPage /> Loading should match snapshot 1`] = `
<h1>
  <div
    class="my-loader"
  >
    <div
      class="my-loader-dots"
    />
  </div>
</h1>
`;
